src/net/p2p.rs

```diff
@@ -3628,12 +3659,54 @@ impl PeerNetwork {
                 }
             };
 
-            // have the downloader request this block if it's new
-            match self.block_downloader {
-                Some(ref mut downloader) => {
-                    downloader.hint_block_sortition_height_available(block_sortition_height, false);
+            let need_block = {
+                if let Ok(Some(sn)) =
+                    SortitionDB::get_block_snapshot_consensus(sortdb.conn(), &consensus_hash)
+                {
+                    if let Ok(inv) = chainstate.get_blocks_inventory(&[(
+                        consensus_hash.clone(),
+                        if sn.sortition {
+                            Some(sn.winning_stacks_block_hash.clone())
+                        } else {
+                            None
+                        },
+                    )]) {
+                        if inv.block_bitvec.len() > 0 {
+                            inv.block_bitvec[0] == 0
+                        } else {
+                            false
```

Comment:
This code assembles a 1-bit inv vector for just this given sortition.  It does exactly one check for the block.

Reply:
Okay, that makes sense.

I have a code-style request here, though -- can you separate the `need_block` and `need_microblock` code blocks into a new function that is somewhat DRY'ed? e.g., adding new:

```rust
fn needs_block(sortition_ch: &ConsensusHash,
               chainstate: &StacksChainState,
               sortdb_conn: &Connection,
               is_microblock: bool)
```

While the method would save some repetition, it would also allow you to avoid the deep if-nesting, because you can just early return the `false` booleans. It might make sense to also emit some warnings (or at least infos or debugs) rather than just squelching the Err variants into `false`.

Reply:
Done